x/genutil/types/genesis.go

```diff
@@ -0,0 +1,157 @@
+	AppVersion    string          `json:"app_version"`
+	GenesisTime   time.Time       `json:"genesis_time"`
+	ChainID       string          `json:"chain_id"`
+	InitialHeight int64           `json:"initial_height"`
+	AppHash       []byte          `json:"app_hash"`
+	AppState      json.RawMessage `json:"app_state,omitempty"`
+
+	// TODO eventually abstract from CometBFT types
+	Validators      []cmttypes.GenesisValidator `json:"validators,omitempty"`
+	ConsensusParams *cmtproto.ConsensusParams   `json:"consensus_params,omitempty"`
+}
+
+// NewAppGenesisWithVersion returns a new AppGenesis with the app name and app version already.
+func NewAppGenesisWithVersion(chainID string, appState json.RawMessage) *AppGenesis {
+	return &AppGenesis{
+		AppName:    version.AppName,
+		AppVersion: version.Version,
+		ChainID:    chainID,
+		AppState:   appState,
+		Validators: nil,
+	}
+}
+
+// ToCometBFTGenesisDoc converts the AppGenesis to a CometBFT GenesisDoc.
+func (ag *AppGenesis) ToCometBFTGenesisDoc() (*cmttypes.GenesisDoc, error) {
+	var consensusParams *cmttypes.ConsensusParams
+	if ag.ConsensusParams != nil {
+		consensusParams = &cmttypes.ConsensusParams{
+			Block: cmttypes.BlockParams{
+				MaxBytes: ag.ConsensusParams.Block.MaxBytes,
+				MaxGas:   ag.ConsensusParams.Block.MaxGas,
+			},
+			Evidence: cmttypes.EvidenceParams{
+				MaxAgeNumBlocks: ag.ConsensusParams.Evidence.MaxAgeNumBlocks,
+				MaxAgeDuration:  ag.ConsensusParams.Evidence.MaxAgeDuration,
+				MaxBytes:        ag.ConsensusParams.Evidence.MaxBytes,
+			},
+			Validator: cmttypes.ValidatorParams{
+				PubKeyTypes: ag.ConsensusParams.Validator.PubKeyTypes,
+			},
+		}
+	}
+
+	return &cmttypes.GenesisDoc{
+		ChainID:         ag.ChainID,
+		InitialHeight:   ag.InitialHeight,
+		AppHash:         ag.AppHash,
+		AppState:        ag.AppState,
+		ConsensusParams: consensusParams,
+		Validators:      ag.Validators,
+	}, nil
+}
+
+// SaveAs is a utility method for saving AppGenesis as a JSON file.
+func (ag *AppGenesis) SaveAs(file string) error {
+	appGenesisBytes, err := json.MarshalIndent(ag, "", "  ")
+	if err != nil {
+		return err
+	}
+
+	return os.WriteFile(file, appGenesisBytes, 0600)
+}
+
+// ValidateAndComplete performs validation and completes the AppGenesis.
+func (ag *AppGenesis) ValidateAndComplete() error {
+	if ag.ChainID == "" {
+		return errors.New("genesis doc must include non-empty chain_id")
+	}
+
+	if len(ag.ChainID) > MaxChainIDLen {
+		return fmt.Errorf("chain_id in genesis doc is too long (max: %d)", MaxChainIDLen)
+	}
+
+	if ag.InitialHeight < 0 {
+		return fmt.Errorf("initial_height cannot be negative (got %v)", ag.InitialHeight)
+	}
+
+	if ag.InitialHeight == 0 {
+		ag.InitialHeight = 1
+	}
+
+	if ag.GenesisTime.IsZero() {
+		ag.GenesisTime = cmttime.Now()
+	}
+
+	// verify that consesus and validators parameters are valid for CometBFT
+	// TODO eventually generalize this for every consensus engine
+	cmtGenesis, err := ag.ToCometBFTGenesisDoc()
```

Comment:
I am not super happy about this part (113 until 124). However, I think it would be premature optimization to generalize it.